provider/pkg/provider/remote/connection.go

```diff
@@ -107,16 +117,73 @@ func (con *Connection) SShConfig() (*ssh.ClientConfig, error) {
 	return config, nil
 }
 
+func (con *Connection) ProxySShConfig() (*ssh.ClientConfig, error) {
+	config := &ssh.ClientConfig{
+		User:            *con.ProxyUser,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	if con.ProxyPrivateKey != nil {
+		signer, err := ssh.ParsePrivateKey([]byte(*con.ProxyPrivateKey))
+		if err != nil {
+			return nil, err
+		}
+		config.Auth = append(config.Auth, ssh.PublicKeys(signer))
+	}
+	if con.ProxyPassword != nil {
+		config.Auth = append(config.Auth, ssh.Password(*con.ProxyPassword))
+		config.Auth = append(config.Auth, ssh.KeyboardInteractive(func(user, instruction string, questions []string, echos []bool) (answers []string, err error) {
+			for i := range questions {
+				answers[i] = *con.ProxyPassword
+			}
+			return answers, err
+		}))
+	}
+
+	return config, nil
+}
+
 // Dial a ssh client connection from a ssh client configuration, retrying as necessary.
-func (con *Connection) Dial(ctx p.Context, config *ssh.ClientConfig) (*ssh.Client, error) {
+func (con *Connection) Dial(ctx p.Context, proxyConfig *ssh.ClientConfig, config *ssh.ClientConfig) (*ssh.Client, error) {
 	var client *ssh.Client
+	var proxyClient *ssh.Client
 	var err error
 	var dialErrorLimit = con.getDialErrorLimit()
 	_, _, err = retry.Until(ctx, retry.Acceptor{
 		Accept: func(try int, nextRetryTime time.Duration) (bool, interface{}, error) {
 			client, err = ssh.Dial("tcp",
 				net.JoinHostPort(*con.Host, fmt.Sprintf("%.0f", *con.Port)),
 				config)
+			if *con.ProxyHost != "" {
```

Comment:
Does it make sense to only construct the proxyConfig here instead of passing it in?

Reply:
Refactored as suggested